swarm/src/connection/handler_wrapper.rs

```diff
@@ -307,27 +307,57 @@ where
             Error<TProtoHandler::Error>,
         >,
     > {
-        while let Poll::Ready(Some((user_data, res))) = self.negotiating_in.poll_next_unpin(cx) {
-            match res {
-                Ok(upgrade) => self
-                    .handler
-                    .inject_fully_negotiated_inbound(upgrade, user_data),
-                Err(err) => self.handler.inject_listen_upgrade_error(user_data, err),
+        loop {
+            // Poll the [`ConnectionHandler`].
+            if let Poll::Ready(res) = self.handler.poll(cx) {
+                match res {
+                    ConnectionHandlerEvent::Custom(event) => {
+                        return Poll::Ready(Ok(Event::Custom(event)));
+                    }
+                    ConnectionHandlerEvent::OutboundSubstreamRequest { protocol } => {
+                        let id = self.unique_dial_upgrade_id;
+                        let timeout = *protocol.timeout();
+                        self.unique_dial_upgrade_id += 1;
+                        let (upgrade, info) = protocol.into_upgrade();
+                        self.queued_dial_upgrades.push((id, SendWrapper(upgrade)));
+                        return Poll::Ready(Ok(Event::OutboundSubstreamRequest((
+                            id, info, timeout,
+                        ))));
+                    }
+                    ConnectionHandlerEvent::Close(err) => return Poll::Ready(Err(err.into())),
```

Comment:
It is not immediately clear that all these branches will `return` because of the noise of constructing the return value. We could extract this into a separate function (`self.poll_handler(cx)`) and only have 1 `return` statement to make this more obvious. What do you think?

Reply:
Yeah that is better :)